Scrape to archive

ondenman opened the pull request:
An election has recently been held for this legislature ([29/10/2016](http://www.ipu.org/parline-e/Modlist.asp)).

The data is currently scraped from Wikipedia but the Wikipedia pages have not been updated for the new term. This PR uses http://www.althingi.is as the source for the new term.

The purpose of this PR is to visit each member page belonging to members listed on the Congressmen, Substitutes and Alternates pages so that each page is saved to the archive.

## [Scraper Change checklist](https://github.com/everypolitician/everypolitician/wiki/Scraper-Change-checklist)
* [x] 1. scraper is on Morph.io under the "everypolitician-scrapers" group?
* [x] 2. scraper's GitHub "Website" link points at morph.io page?
* [x] 3. scraper is set to auto-run?
* [x] 4. scraper is archiving?
* [x] 5. legislature has a scraper webhook set?

ondenman committed:
Add MembersPage

Represent a page listing members of the legislature. Exposes a list
of member urls.

ondenman committed:
Archive member pages

ondenman committed:
Add IcelandicMembersPage

Represents a page on which members are listed.

ondenman committed:
Archive Icelandic member pages from Members, Alternates and Substitutes pages

The Icelandic version of the site covers alternates and substitutes
as well as the members listed on the English page.

ondenman committed:
Rename MembersPage to EnglishMembersPage

Since we now have IcelandicMembersPage, I have renamed MembersPage
to ENglishMembersPage to prevent any confusion that this might
be a base class. This class specifically represents a page listing
the members of the legislature in English.

ondenman committed:
Scrape both English and Icelandic pages

ondenman committed:
Tidy for Rubocop

ondenman committed:
Add rake test task

ondenman committed:
Configure Travis to run tests

ondenman committed:
Add minitest gem

ondenman committed:
Bundle install

ondenman committed:
Add tests

tmtmtmtm reviewed (changes requested):
This is coming together nicely, but running tests against live URLs isn't a great idea. I'd suggest recording these with VCR.

ondenman committed:
Add VCR gems

ondenman committed:
Bundle install

ondenman committed:
Add test helper

ondenman committed:
Add vcr to tests

tmtmtmtm reviewed (changes requested):
_(NB: I've squashed a bunch of the latest commits down to make the flow simpler here. In particular, it would be good if you rubocop-tidied each file as you go (or at least fix them up as if you had before pushing) rather than having one big "tidy up" commit at the end.)_

You seem to have forgotten to check in the VCR cassettes…

tmtmtmtm commented on tests/icelandic_member_page_test.rb:
```diff
@@ -0,0 +1,27 @@
+    end
+  end
+
+  describe 'member_urls' do
```
Same here with two `describes` with the same name

tmtmtmtm commented on .travis.yml:
```diff
@@ -1,5 +1,6 @@
 language: ruby
 script:
   - bundle exec rake rubocop
+  - bundle exec rake test
```
might be better to have the default rake task be both `rubocop` and `test` and then you could drop this section entirely.

tmtmtmtm commented on scraper.rb:
```diff
@@ -0,0 +1,30 @@
+      .member_urls
+      .each do |member_url|
+      _ = scraped_page_for(member_url)
+    end
```
there's some quite odd formatting going on here. 

```ruby
scraped_page_for(page_url, key).member_urls.each do |member_url|
```
all fits merrily onto a single line, though, so I don't think there's much to be gained (and some confusion to be had), but splitting it across multiple lines like this.

tmtmtmtm commented on tests/english_member_page_test.rb:
```diff
@@ -0,0 +1,24 @@
+describe EnglishMembersPage, :vcr do
+  describe 'member_urls' do
+    it 'should return the expected number of urls' do
+      url = 'http://www.althingi.is/altext/cv/en/'
```
if this URL is going to be used in every test, probably better to hoist it out.

tmtmtmtm commented on tests/english_member_page_test.rb:
```diff
@@ -0,0 +1,24 @@
+  describe 'member_urls' do
+    it 'should return the expected number of urls' do
+      url = 'http://www.althingi.is/altext/cv/en/'
+      EnglishMembersPage.new(response: Scraped::Request.new(url: url).response)
```
same here — if every test is going to use the same Page, better to move it out, and keep the body of the test for things that will vary.

tmtmtmtm commented on tests/english_member_page_test.rb:
```diff
@@ -0,0 +1,24 @@
+    end
+  end
+
+  describe 'member_urls' do
```
You have two `describe` blocks with the same name. If these need to be separate, you should give them different names, otherwise you should simply have separate `it` blocks within the one `describe` section.

tmtmtmtm replied:
You still have two describe blocks with the same name in both files.

tmtmtmtm commented on tests/icelandic_member_page_test.rb:
```diff
@@ -0,0 +1,27 @@
+  describe 'member_urls' do
+    it 'should return the expected number of urls' do
+      url = 'http://www.althingi.is/thingmenn/althingismenn/'
+      IcelandicMembersPage.new(response: Scraped::Request.new(url: url)
```
same as above with hoisting these.

chrismytton commented on Gemfile:
```diff
@@ -6,6 +6,8 @@ ruby '2.3.3'
 git_source(:github) { |repo_name| "https://github.com/#{repo_name}.git" }
 
+gem 'minitest'
+gem 'minitest-vcr'
```
Based on past experience I think it's probably best to use `minitest-around` rather than `minitest-vcr`. There are two reasons for this:

1. Using `minitest-around` allows you to use a single cassette across multiple `it` blocks. In this pull request there are currently four cassettes, but there are only two urls being hit so it only needs two cassettes.
2. `minitest-around` decouples your test names from your cassettes, so renaming a test doesn't cause your cassette to silently be re-recorded.

You can see `minitest-around` in action [in `wikisnakker`](https://github.com/everypolitician/wikisnakker/blob/d2bd31d9ac1ff10cf6498f0072523525d7b548ac/test/single_test.rb#L4).

ondenman committed:
Make both Rubocop and Test tasks default

ondenman committed:
Add minitest-around gem

minitest-around allows the use of a single vcr cassette over
multiple it blocks and de-couples test names from
cassettes.

ondenman committed:
Bundle install

ondenman committed:
Use minitest-around

tmtmtmtm committed:
Simplify 'around' blocks

tmtmtmtm commented on tests/english_member_page_test.rb:
```diff
@@ -0,0 +1,31 @@
+    VCR.use_cassette('EnglishMembersPage') do
+      test.call
+    end
+  end
```
This is a bit of a mouthful. This could be simplified to:

```ruby
  around { |test| VCR.use_cassette('EnglishMembersPage', &test) }
```

tmtmtmtm commented on tests/english_member_page_test.rb:
```diff
@@ -0,0 +1,31 @@
+
+  describe 'member_urls' do
+    it 'should return an expected url in its result' do
+      url = 'http://www.althingi.is/altext/cv/en/'
```
Why shadow the outside `url` here?

tmtmtmtm commented on tests/english_member_page_test.rb:
```diff
@@ -0,0 +1,31 @@
+
+  describe 'member_urls' do
+    it 'should return the expected number of urls' do
+      EnglishMembersPage.new(response: Scraped::Request.new(url: url).response)
```
My suggestion was to hoist _this_, not just to the `url`

tmtmtmtm commented on tests/english_member_page_test.rb:
```diff
@@ -0,0 +1,31 @@
+require_relative '../lib/english_members_page.rb'
+
+describe EnglishMembersPage do
+  url = 'http://www.althingi.is/altext/cv/en/'
```
Better to do this with a `let` (or ideally a `subject`)

tmtmtmtm committed:
Extract `subject` responses

tmtmtmtm committed:
Group tests together

Don't have multiple blocks with the same name

tmtmtmtm committed:
Use default travis task

No need to tell travis to run `bundle exec rake` — that's what it does
by default:
 https://docs.travis-ci.com/user/languages/ruby/#Default-Test-Script

tmtmtmtm commented on .travis.yml:
```diff
@@ -1,5 +1,5 @@
 language: ruby
 script:
-  - bundle exec rake rubocop
+  - bundle exec rake
```
There's no need to tell travis to run that — that's what it does by default: https://docs.travis-ci.com/user/languages/ruby/#Default-Test-Script

tmtmtmtm reviewed (commented):
I've pushed some additional commits to tidy up these problems.

tmtmtmtm committed:
Use AbsoluteUrls decorator

tmtmtmtm commented on lib/english_members_page.rb:
```diff
@@ -0,0 +1,10 @@
+class EnglishMembersPage < Scraped::HTML
+  field :member_urls do
+    noko.xpath('//table[@id="members"]/tbody/tr/td/b/a/@href').map do |row|
+      URI.join(url, row.text).to_s
```
The built-in AbsoluteUrls decorator will do this for you: I've pushed a commit showing that.
